ActionInfo[] action_info_list